enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,745 @@
+
+
+## Support Procedures
+
+To find out any configuration issues:
+- Documentation of edge cases and potential pitfalls discovered during implementation.
+- `microshift doctor low-latency` command to verify if the pieces involved in tuning host for low latency
+  are as expected according to developers' knowledge. Mainly comparing values between different
+  config files, verifying that RT kernel is installed and booted, tuned profile is active, etc.
+
+To discover any performance issues not related to missing configuration:
+- Adapting some parts of OpenShift documentation.
+- Referring user to [Red Hat Enterprise Linux for Real Time](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/9) documentation.
+
+## Alternatives
+
+### Deploying Node Tuning Operator
+
+Most of the functionality discussed in scope of this enhancement is already handled by Node Tuning
+Operator (NTO). However incorporating it in the MicroShift is not the best way for couple reasons:
+- NTO depends on Machine Config Operator which also is not supported on MicroShift,
+- MicroShift takes different approach to host management than OpenShift,
+- MicroShift being intended for edge devices aims to reduce runtime resource consumption and
+  introducing operator is against this goal.
+
+
+### Reusing NTO code
+
+Instead of deploying NTO, its code could be partially incorporated in the MicroShift.
+However this doesn't improve the operational aspects: MicroShift would transform a CR into TuneD,
+CRI-O config, and kubelet configuration, which means it's still a controller, just running in
+different binary and that doesn't help with runtime resource consumption.
+
+Parts that depend on the MCO would need to be rewritten and maintained.
+
+Other aspect is that NTO is highly generic, supporting many configuration options to mix and match
+by the users, but this enhancement focuses solely on Low Latency.
+
+
+### Providing users with upstream documentations on how to use TuneD and configure CRI-O 
+
+This is least UX friendly way of providing the functionality.
+Responsibility of dev team is to remove common hurdles from user's path so they make less mistakes
+and want to continue using the product.
+
+
+### Applying user requested TuneD profile on every start of the `microshift-tuned.service`
+
+This was preferred option before it was discovered that when profile is applied, tuned will
+first remove all the old kernel args, then append new kernel args resulting in creation of new staged OSTree deployment.
+This happens even if the the same profile is being reapplied.
+
+
+### Using Kubelet's drop-in config directory feature
+
+This is alternative to "MicroShift config -> Kubelet config" passthrough.
+Using drop-ins would mean there is less implementation on MicroShift side, but it also gives away
+full power of Kubelet config. With passthrough we can later introduce some validation or
```

Comment:
we already expose the drop-in for CRIO , IMO we should do use the same for kubelet, its  less cumbersome then implementing pass-through.

Reply:
I think the semantics are a bit different. We don't expose CRIO's configuration, we don't manage it, sure, we add some configs, but it's not that microshift packages crio.

With kubelet it's a bit different because it's an integral part of microshift.
When I brought this topic up this week, @DanielFroehlich mentioned "reducing scope of the configuration" - with a drop-in we're giving away full power; with passthrough we can introduce some control - limit settings available or provide some validation

Reply:
okay, so we embed the kubelet configuration and dont expose it on the disk.

Reply:
@eslutsky - is this resolved?